File: 0874-walking-robot-simulation/0874-walking-robot-simulation.cpp
class Solution {
public:
    int robotSim(vector<int>& commands, vector<vector<int>>& obstacles) {
        int res = 0;
        int x = 0;
        int y = 0;
        int curDir = 0;
        unordered_set<int> obs;
        vector<pair<int, int>> dirs = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};
        for (int i = 0; i < obstacles.size(); i++) {
            obs.insert(obstacles[i][0] * 60013 + obstacles[i][1]);
        }
        for (int i = 0; i < commands.size(); i++) {
            if (commands[i] == -1) {
                curDir++;
                curDir = ((curDir % 4) + 4) % 4;
                continue;
            } else if (commands[i] == -2) {
                curDir--;
                curDir = ((curDir % 4) + 4) % 4;
                cout << curDir << endl;
                continue;
            }
            for(int j = 0; j < commands[i]; j++)
            {
                if(obs.count((x + dirs[curDir].first) * 60013 + (y + dirs[curDir].second)))
                {
                    break;
                }
                x += dirs[curDir].first;
                y += dirs[curDir].second;
                res = max(res, x * x + y * y);
            }
        }
        return res;
    }
};

File: 0874-walking-robot-simulation/0874-walking-robot-simulation.py
class Solution:
    def robotSim(self, commands: List[int], obstacles: List[List[int]]) -> int:
        res = 0
        x = 0
        y = 0
        curDir = 0
        obs = {obX + obY * 60013 for obX, obY in obstacles}
        dirs = [(0, 1), (1, 0), (0, -1), (-1, 0)]
        for c in commands:
            if c == -1:
                curDir += 1
                curDir %= 4
                continue
            elif c == -2:
                curDir -= 1
                curDir %= 4
                continue
            for j in range(c):
                if ((x + dirs[curDir][0]) + 60013 * (y + dirs[curDir][1])) in obs:
                    break
                x += dirs[curDir][0]
                y += dirs[curDir][1]
                res = max(res, x * x + y * y)
        return res
